replay2/memory_remapper/memory_remapper.cc

```diff
@@ -14,4 +14,98 @@
 
 #include "memory_remapper.h"
 
-void foo(int var) { return; }
+namespace agi {
+namespace replay2 {
+
+	ReplayAddress MemoryRemapper::AddMapping(const MemoryObservation& observation)
+	{
+		const CaptureAddress captureAddress = observation.captureAddress();
+
+		auto iter = CaptureAddressRangeIter(captureAddress);
+		if(iter != captureAddressRanges_.end()) {
+
+			const CaptureAddressRange captureAddressRange = iter->first;
+
+			const intptr_t offset = captureAddress.charPtr() -captureAddressRange.baseAddress().charPtr();
+			if(offset < captureAddressRange.length()) {
+				throw AddressAlreadyMappedException();
+			}
+		}
+
+		const size_t mappingLength = observation.resourceGenerator()->length();
+
+		char* replayAllocation = new char [mappingLength];
+		const ReplayAddress replayAddress(replayAllocation);
+
+		observation.resourceGenerator()->generate(replayAddress);
+
+		const CaptureAddressRange captureAddressRange = CaptureAddressRange(captureAddress, mappingLength);
+		const ReplayAddressRange replayAddressRange = ReplayAddressRange(replayAddress, mappingLength);
+
+		const auto newMapping = std::make_pair(captureAddressRange, replayAddressRange);
+		captureAddressRanges_.insert(newMapping);
```

Comment:
Can do. I thought it "read" nicer this way, but I can change if you prefer it.

Reply:
I would prefer because I think this also have an extra copy. I remember that vectors has this `push_back` vs. `emplace_back` which if I remember people would use `emplace` because it had one less copy. Are there any similar thing for maps?

These are not very big performance concerns but if we make using them habit, we wouldn't need to change in the future everywhere and it would be clear for people to follow the style. What do you think about it?

Reply:
Changed

Reply:
As a general comment though, it's good to be aware of this stuff when you're using C++, but I mostly lean towards most readable code and don't concern myself with removing every small object copy outside of hot loops. Compilers are surprisingly good at eliminating copies these days, and we can always hand optimise when we see a codepath has become a bottleneck.